returns/interfaces/filterable.py

```diff
@@ -0,0 +1,41 @@
+from abc import abstractmethod
+from typing import Callable, Generic, TypeVar
+
+from returns.primitives.hkt import Kind1
+
+_InnerType = TypeVar('_InnerType')
+
+_FilterableType = TypeVar('_FilterableType', bound='Filterable')
+
+
+class Filterable(Generic[_InnerType]):
+    """
+    Represents container that can apply filter over inner value.
+
+    There are no aliases or ``FilterableN` for ``Filterable`` interface.
+    Because it always uses one type.
+
+    Not all types can be ``Filterable`` because we require
+    a possibility to access internal value and to model a case,
+    where the predicate is false
+
+    .. code:: python
+
+        >>> from returns.maybe import Nothing, Some
+        >>> from returns.pointfree import filter_
+
+        >>> def example(argument: int) -> bool:
+        ...     return argument % 2 == 0
+
+        >>> assert filter_(example)(Some(5)) == Nothing
+        >>> assert filter_(example)(Some(6)) == Some(6)
+        >>> assert filter_(example)(Nothing) == Nothing
+
+    """
+
+    @abstractmethod
+    def filter(
```

Comment:
Future note: we can also make an `overload` for `TypeGuard` case.
See https://github.com/python/typeshed/pull/6140

Reply:
Could you elaborate a bit what can we do with type guard here? We are not filtering for example a list with different types. Filter executed on `_FilterableType` always returns `_FilterableType`. Did I miss something?

Reply:
Nevermind, this was a self note 🙂 
This is even not supported by mypy yet.

Reply:
It looks like mypy [supports](https://mypy.readthedocs.io/en/latest/type_narrowing.html#user-defined-type-guards) `TypeGuard` now 🎉 :